### ISSUE TITLE
[API Proposal]: Add `IReadOnlyList<TElement>` to System.Linq.Grouping<TKey, TElement>

### ISSUE THREAD

**Issue by MatthewSteeples** [dotnet/runtime#92165]
### Background and motivation

Related to https://github.com/dotnet/runtime/issues/86219
Similar to https://github.com/dotnet/runtime/issues/91756 and https://github.com/dotnet/runtime/issues/31001

This class already implements the functionality of the interface without explicitly implementing it (as it implements `IList<TElement>`).

The indexer on `Lookup<TKey, TElement>` returns this as an `IEnumerable<TElement>` when using the indexer. We have analysers that ensure that we don't enumerate IEnumerables more than once so we have some code that either casts them to their equivalent collection or converts the `IEnumerable<T>` into an `ICollection<T>` or `IReadOnlyCollection<T>`. In this case we end up passing the Grouping in as a parameter to ReadOnlyCollection because it doesn't explicitly implement the interface.

```csharp
public static IReadOnlyCollection<T> AsReadOnlyCollection<T>(this IEnumerable<T> values)
{
	if (values is IReadOnlyCollection<T> readonlyCollection)
		return readonlyCollection; //Ideally we'd want it to return here

	if (values is IList<T> list)
	{
		if (list.Count == 0)
			return Array.Empty<T>();

		return new ReadOnlyCollection<T>(list); //It actually returns here
	}

        ///rest of implementation
}

### API Proposal

Replace

```csharp
public class Grouping<TKey, TElement> : IGrouping<TKey, TElement>, IList<TElement>
```

with

```csharp
public class Grouping<TKey, TElement> : IGrouping<TKey, TElement>, IList<TElement>, IReadOnlyList<TElement>
```

### API Usage

No change in usage of the class

### Alternative Designs

Because making `IList<T>` implement `IReadOnlyList<T>` will cause breaking changes it has been recommended to update the individual classes

### Risks

Potential performance penalty of implementing an additional interface

**Comment by eiriktsarpalis** [comment]
It is something that we do for other public collection types as well, so I don't see why we couldn't do it here. @stephentoub any objections?

Note that we might also want to update a few internal types that also implement the `IGrouping` interface for consistency:

https://github.com/dotnet/runtime/blob/197268373bcf990756561fa728f971dc30332012/src/libraries/System.Linq.Parallel/src/System/Linq/Parallel/QueryOperators/Unary/GroupByQueryOperator.cs#L29-L30

**Comment by stephentoub** [comment]
> @stephentoub any objections?

I don't have a problem with relevant internal LINQ types implementing IROL as an implementation detail. There's the ever present discussion about whether type checks in LINQ should _query_ for IROL, which is a separate consideration.

**Comment by MatthewSteeples** [comment]
Agreed. The checking for IROL (or IROC in our case) was only meant as a sample of how we're using it, not a suggestion to update LINQ

**Comment by stephentoub** [comment]
There's one potential issue to be aware of, ensure we have tests for, and if it does manifest, decide what to do about. Historically we've had issues where folks complained about changes that resulted in side effects from enumeration not occurring, eg if we made the type returned from a select on an array implement IROC, then code that was either querying for the interface and accessing its Count or else enumerating to get the count would no longer enumerate, in which case the delegate to Select wouldn't be invoked where it was previously.

**Comment by eiriktsarpalis** [comment]
> I don't have a problem with relevant internal LINQ types implementing IROL as an implementation detail.

It would also be adding it to a public class, `Grouping<TKey, TElement>`.

**Comment by stephentoub** [comment]
> > I don't have a problem with relevant internal LINQ types implementing IROL as an implementation detail.
> 
> It would also be adding it to a public class, `Grouping<TKey, TElement>`.

Seems fine. Just looked at it... we should delete the comments on the type, which at this point are silly.

**Comment by bartonjs** [comment]
This issue prompted two competing options: explore trying to use a DIM to make `IList<T> : IReadOnlyList<T>`, or writing a tool (analyzer?) to identify all IList-implementing types that are missing IReadOnlyList and fixing them in bulk.

So it's in principle approved, but not quite yet pending followups.

@terrajobst

**Comment by Mrxx99** [comment]
> use a DIM to make `IList<T> : IReadOnlyList<T>`

This would solve so many pain points
